content/inside-rust/test-infra-mar-2025.md

```diff
@@ -0,0 +1,264 @@
+
+- **Precise**: `//~?` annotations will not match against diagnostics that do have line information.
+- **Exhaustive**: if multiple diagnostics lacking line information in the main test file are present, then all of them need to be annotated with `//~?`.
+
+Combined, `//~?` helps to prevent accidentally blessing away diagnostics that don't have associated line information on the main test file.
+
+`//@ error-pattern` are still useful for e.g. runtime error messages that do not have specific spans, or compile time messages that require imprecise matching due to multiline platform-specific diagnostics. `//~?` should be preferred where suitable.
+
+[pr-138865]: https://github.com/rust-lang/rust/pull/138865
+
+
+## Notable changes
+
+This section is intended to be like "compatibility notes" but for human test writers.
+
+### In preparation of stage 0 std redesign, it is now possible to use stage 0 libtest for `compiletest`
+
+The [stage 0 std redesign](https://github.com/rust-lang/rust/pull/119899) means that building the library tree would require a stage 1 compiler. `compiletest` is somewhat strange in that it currently depends on in-tree libtest -- meaning that with the stage 0 std redesign, changes to the compiler tree would necessitate rebuilding `compiletest`, causing more friction for compiler development cycle. To mitigate this, bootstrap [now supports a new configuration option to switch `compiletest` to depend on stage 0 libtest instead of in-tree libtest][pr-139386], so that changes to `compiler/` tree would not cause `compiletest` to rebuild.
+
+```toml
+[build]
+compiletest-use-stage0-libtest = true
+```
+
+Note that as a trade-off, we do check this force-stage0-libtest configuration in PR CI. This means that there are possible windows of time where libtest may be modified before bumping stage 0 compiler, which may require the PR author modifying libtest to add `cfg(bootstrap)`/`cfg(not(bootstrap))` to compiletest bits depending libtest programmatic API as suitable. This is done under the assumption that libtest programmatic API changes are comparatively much less frequent. If this proves to be a significant burden, the bootstrap team would be open to revisiting this scheme.
+
+Note in over the long term, we'd like to migrate away from depending on in-tree libtest (both to render this hack unnecessary and to have more control over test execution).
+
+[pr-139386]: https://github.com/rust-lang/rust/pull/139386
+
+### `bootstrap.toml` is now also accepted as bootstrap configuration file under checkout root
+
+Previously, [bootstrap] only accepted `config.toml` as a configuration file under checkout root. This is potentially confusing since Cargo also uses `config.toml` as its configuration file. Now, [bootstrap will also accept `bootstrap.toml` as the configuration file][pr-137081]. The example config file `config.example.toml` has also been renamed to `bootstrap.example.toml`. `config.toml` will continue to work.
```

Comment:
>  Now, [bootstrap will also accept `bootstrap.toml` as the configuration file][pr-137081]

This isn't fully correct. Bootstrap renamed `config.toml` to `bootstrap.toml`, but it will also support `config.toml` for a while to avoid BC on different checkouts. So `config.toml` won't work forever; it's just a temporary fallback.

Reply:
Right. Let me word this as "`config.toml` has been renamed `bootstrap.toml`" but `config.toml` is currently accepted as a temporary fallback to aid migration, but will eventually be rejected.

Reply:
Additionally, I can also move this bit out of *this* blog post, and go with the stage 0 redesign blog post. What do you think?

Reply:
In any case, fixed the wording to reflect that `config.toml` is a temporary churn-mitigation fallback, but will be phased out eventually.